### PR TITLE
Update EventHubs to enable live testing in sovereign clouds for multiple services

### PR DESCRIPTION
**Pull request by v-hongli1**
These changes enable EventHubs to run live tests against Public, UsGov and China.

1. Update file : test-resources.json
Match with the `ArmTemplateParameter` setting, so change the parameter name of the endpoint suffix to `serviceBusEndpointSuffix`.

2. Update file: EventHubAsyncClientIntegrationTest.java
The way to get the Credential parameter is changed to get from `ConnectionStringProperties`. 

3. Update file: EventHubConsumerAsyncClientIntegrationTest.java
Fixed `Expected messages to be received from all partitions. There are: 1 ==> expected: <true> but was: <false>` Exception, so change the verification parameter settings for the test case `EventHubConsumerAsyncClientIntegrationTest.receivesMultiplePartitions`. 

4. Update files: EventHubClientBuilderTest.java, EventHubConsumerAsyncClientTest.java, EventHubConsumerClientTest.java, EventHubProducerAsyncClientTest.java, EventHubSharedKeyCredentialTest.java, EventProcessorClientBuilderTest.java, EventHubReactorConnectionTest.java, EventHubConnectionStringPropertiesTest.java
Match with the `ArmTemplateParameter` setting, so change the way of get endpoint suffix to `Configuration.getGlobalConfiguration().get("AZURE_EVENTHUBS_ENDPOINT_SUFFIX", ".servicebus.windows.net")`.

5. Update file: IntegrationTestBase.java
Fixed the way for generate signature of `EventHubs` connection string and get value of configuration.

6. Update file: EventPositionIntegrationTest.java
In order to dispose `EventHubConsumerAsyncClient` after each test case, fixed the method `EventPositionIntegrationTest.afterTest`.

Pipeline results:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1172532&view=results

@benbp, @scottaddie, @joshfree, @lilyjma, @ramya-rao-a and @anuchandy for notification.

### CODE REVIEW

**Comment by v-xuto** [comment]
@benbp, @scottaddie, @joshfree, @lilyjma, @ramya-rao-a and @anuchandy  Could you please review this PR?

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java [hunk 1]
```diff
@@ -151,17 +151,18 @@ void getPropertiesWithCredentials() {
...
+        String eventHubName = properties.getEntityPath();
 
         final EventData testData = new EventData(TEST_CONTENTS.getBytes(UTF_8));
 
         EventHubProducerAsyncClient asyncProducerClient = new EventHubClientBuilder()
-                .credential(fullyQualifiedNamespace, eventHubName,
-                        new AzureNamedKeyCredential(sharedAccessKeyName, sharedAccessKey))
-                .buildAsyncProducerClient();
+            .credential(fullyQualifiedNamespace, eventHubName,
+                new EventHubSharedKeyCredential(sharedAccessKeyName,
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java:163]: Is there a reason we're updating this test that should test AzureNamedKeyCredential to EventHubSharedKeyCredential?

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java [hunk 2]
```diff
@@ -180,15 +181,15 @@ public void sendAndReceiveEventByAzureSasCredential() {
...
         String sharedAccessSignature = properties.getSharedAccessSignature();
-        String eventHubName = getEventHubName();
+        String eventHubName = properties.getEntityPath();
 
         final EventData testData = new EventData(TEST_CONTENTS.getBytes(UTF_8));
 
         EventHubProducerAsyncClient asyncProducerClient = new EventHubClientBuilder()
                 .credential(fullyQualifiedNamespace, eventHubName,
-                        new AzureSasCredential(sharedAccessSignature))
+                        new EventHubSharedKeyCredential(sharedAccessSignature))
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java:192]: This test case is to test a SasCredential.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 3]
```diff
@@ -439,30 +439,43 @@ public void receivesMultiplePartitions() {
             () -> logger.info("Completed"));
 
         // Act & Assert
+        AtomicInteger removedPartitionsNumber = new AtomicInteger(expectedNumber);
         try {
             StepVerifier.create(consumer.receive(false)
                 .filter(x -> TestUtils.isMatchingEvent(x.getData(), MESSAGE_TRACKING_ID))
                 .take(expectedNumber))
-                .assertNext(event -> assertPartitionEvent(event, producer.getEventHubName(), allPartitions,
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:446]: The intention of this test was that we would receive from all 6 partitions. The test here is updated so that it receives 6 times, it doesn't discern between what partitions were received from.

Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:446]: Since this test case sets a limit on the number of received pieces, in order to eliminate the interference of other test case data on the verification results, so modified the filter conditions of the receive event data.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java (hunk 2 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java]: The code has been rollback to the original settings and successfully passed the test on UsGov and China cloud by pipeline.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java (hunk 1 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubAsyncClientIntegrationTest.java]: The code has been rollback to the original settings and successfully passed the test on UsGov and China cloud by pipeline.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java (hunk 3 above)
Comment (v-xuto) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:446]: @conniey Any ideas?

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR. If there is no problem, please approve this PR.

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR when you are free. If there is no problem, please approve this PR.

**Comment by ramya-rao-a** [comment]
@v-xuto Can you please resolve the merge conflicts here?

**Comment by v-xuto** [comment]
@ramya-rao-a I have resolved the merge conflicts here.

**Comment by ramya-rao-a** [comment]
Thanks @v-xuto 
Let's wait for @conniey to review

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR when you are free. If there is no problem, please approve this PR.

**Comment by v-xuto** [comment]
@anuchandy Could you please help review this PR?

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 4]
```diff
@@ -423,9 +423,10 @@ public void receivesMultiplePartitions() {
 
         // This is the one we'll mutate.
         final Set<Integer> expectedPartitions = new HashSet<>(allPartitions);
+        final Set<Integer> matchPartitions = Collections.synchronizedSet(new HashSet<>(allPartitions));
         final int expectedNumber = 6;
 
-        Assumptions.assumeTrue(expectedPartitions.size() <= expectedNumber,
+        Assumptions.assumeTrue(expectedPartitions.size() == expectedNumber,
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:429]: Is there a reason we updated this condition? It doesn't match the error message. (Though, the error message is incorrect and should at least be updated to: "Cannot run this test if there are less partitions than expected.")

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 5]
```diff
@@ -423,9 +423,10 @@ public void receivesMultiplePartitions() {
 
         // This is the one we'll mutate.
         final Set<Integer> expectedPartitions = new HashSet<>(allPartitions);
+        final Set<Integer> matchPartitions = Collections.synchronizedSet(new HashSet<>(allPartitions));
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:426]: Is there a reason why we create another variable? `matchPartitions` is the same as `allPartitions`

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 6]
```diff
@@ -441,7 +442,8 @@ public void receivesMultiplePartitions() {
         // Act & Assert
         try {
             StepVerifier.create(consumer.receive(false)
-                .filter(x -> TestUtils.isMatchingEvent(x.getData(), MESSAGE_TRACKING_ID))
+                .filter(x -> TestUtils.isMatchingEvent(x.getData(), MESSAGE_TRACKING_ID)
+                    && TestUtils.isMatchPartition(x.getPartitionContext().getPartitionId(), matchPartitions))
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:446]: There are two sets being mutated here `matchPartitions` and `expectedPartitions`. We don't need TestUtils.isMatchPartition either, that is asserted in assertPartitionEvent. 

You can update this to:

```java
&& expectedPartitions.contains(x.getPartitionContext().getPartitionId()))
```

since `expectedPartitions` is the set that we are still waiting on data from.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java [hunk 7]
```diff
@@ -128,17 +137,55 @@ static String getConnectionString() {
     }
 
     static String getConnectionString(boolean withSas) {
+        String connectionString = GLOBAL_CONFIGURATION.get(EVENT_HUB_CONNECTION_STRING_ENV_NAME);
         if (withSas) {
-            return System.getenv(EVENT_HUB_CONNECTION_STRING_WITH_SAS);
+            final String shareAccessSignatureFormat = "SharedAccessSignature sr=%s&sig=%s&se=%s&skn=%s";
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java:142]: Is there a reason we are calculating the SAS here? The intention was that if the test environment variable had a SAS key in it, we'd run that version of the test.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/TestUtils.java [hunk 8]
```diff
@@ -157,6 +158,14 @@ public static boolean isMatchingEvent(EventData event, String expectedValue) {
             && expectedValue.equals(event.getProperties().get(MESSAGE_ID));
     }
 
+    public static boolean isMatchPartition(String partitionId, Set<Integer> matchPartitions) {
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/TestUtils.java:161]: This isn't needed.

File: sdk/keyvault/test-resources.json [hunk 9]
```diff
@@ -206,53 +206,53 @@
...
-                "recover",
-                "restore",
-                "rotate",
-                "sign",
-                "setrotationpolicy",
-                "unwrapKey",
-                "update",
-                "verify",
-                "wrapKey"
+                "Get",
```
Comment (conniey) [inline sdk/keyvault/test-resources.json:209]: We shouldn't be modifying content in keyvault. Or this should be in another PR.

**Review by conniey — CHANGES_REQUESTED** [review]

File: sdk/keyvault/test-resources.json (hunk 9 above)
Comment (hongli750210) [reply sdk/keyvault/test-resources.json]: Rollbacked unnecessary fixes

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/TestUtils.java (hunk 8 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/TestUtils.java:161]: Removed the unwanted method

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java (hunk 7 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java:142]: The test case  `EventHubAsyncClientIntegrationTest.sendAndReceiveEventByAzureSasCredential` and `EventHubProducerAsyncClientIntegrationTest.sendWithSasConnectionString` need to use Sas connection, but the relevant parameter is not recorded in `test-resources.json`, so added the calculation of `SharedAccessSignature`.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java (hunk 6 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:446]: Fixed in the new version.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java (hunk 5 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:426]: Fixed in the new version.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java (hunk 4 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:429]: Fixed in the new version.

**Comment by v-xuto** [comment]
@conniey We have updated by your comments. Please help re-review this PR. Thanks a lot.

**Comment by v-xuto** [comment]
@conniey We have updated by your comments. Please help re-review this PR. Thanks a lot.

**Comment by ramya-rao-a** [comment]
@srnagar Can you help with the review here?

**Comment by ramya-rao-a** [comment]
@v-xuto The analyze step in CI is failing for this PR, can you please take a look?

**Comment by hongli750210** [comment]
@ramya-rao-a We've fixed the Ci error. Thank you for your reminder.

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR. Thanks a lot.

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR. Thanks a lot.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java (hunk 7 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java:142]: I meant that we don't want to duplicate this logic. I typed out what I was hinting at.... we don't want to copy this entire logic and use existing classes to get the shared access signature.

```java
var connectionProperties = new ConnectionStringProperties(connectionString);
var credential = new EventHubSharedKeyCredential(connectionProperties.getSharedAccessKeyName(), connectionProperties.getSharedAccessKey());

var tokenManager = new AzureTokenManagerProvider(
    CbsAuthorizationType.SHARED_ACCESS_SIGNATURE, connectionProperties.getFullyQualifiedNamespace(), ClientConstants.AZURE_ACTIVE_DIRECTORY_SCOPE);

var scopes = tokenManager.getScopesFromResource(resourceUrl);
var tokenRequest = new TokenRequestContext().addScopes(scopes);

// This contains the shared access signature.
Mono<AccessToken> accessTokenMono = credential.getToken(tokenRequest);
AccessToken accessToken = accessTokenMono.block(Duration.ofSeconds(5));

assertNotNull(accessToken);

String sharedAccessSignature = accessToken.getToken();
```

File: sdk/keyvault/test-resources.json [hunk 10]
```diff
@@ -206,53 +206,53 @@
             "objectId": "[parameters('testApplicationOid')]",
```
Comment (conniey) [inline sdk/keyvault/test-resources.json:206]: This is not part of the PR. Please undo this.

File: sdk/eventhubs/test-resources.json [hunk 11]
```diff
@@ -12,9 +12,12 @@
             "type": "string",
             "defaultValue": "javaeventhub"
         },
-        "eventHubNamespaceHostNameSuffix": {
+        "serviceBusEndpointSuffix": {
```
Comment (conniey) [inline sdk/eventhubs/test-resources.json:15]: Why service bus? This should be Event Hubs still.

**Review by conniey — COMMENTED** [review]
I have a couple of comments.

File: sdk/keyvault/test-resources.json (hunk 10 above)
Comment (hongli750210) [reply sdk/keyvault/test-resources.json]: This file has been rebased to the main branch version

File: sdk/eventhubs/test-resources.json (hunk 11 above)
Comment (hongli750210) [reply sdk/eventhubs/test-resources.json:15]: There is no [eventHubNamespaceHostNameSuffix] parameter record in the configuration of the cloud service, and the Endpoint suffix of the EventHubs Service is exactly the same as that of the Service Bus, so the parameter [serviceBusEndpointSuffix] is used directly.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 12]
```diff
@@ -440,22 +437,23 @@ public void receivesMultiplePartitions() {
...
+            thread.start();
+            thread.join(TIMEOUT.toMillis());
+
+            assertEquals(expectedNumber, partitionEvents.size());
+            partitionEvents.stream().forEach(event ->
+                assertPartitionEvent(event, producer.getEventHubName(), allPartitions,
+                    expectedPartitions));
+
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java]: According to your comment, fixed in the new version.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java (hunk 7 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/IntegrationTestBase.java]: According to your comment, fixed in the new version.

**Comment by v-xuto** [comment]
@conniey Please help re-review this PR. Thanks a lot.

**Comment by v-xuto** [comment]
@conniey We have updated by your comments. Please help re-review this PR. Thanks a lot.

**Comment by v-xuto** [comment]
@conniey We have updated by your comments. Please help re-review this PR. Thanks a lot.

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java [hunk 13]
```diff
@@ -21,13 +21,7 @@
...
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java:24]: Our check styles don't allow for * improts.

**Review by conniey — APPROVED** [review]

File: sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java (hunk 13 above)
Comment (hongli750210) [reply sdk/eventhubs/azure-messaging-eventhubs/src/test/java/com/azure/messaging/eventhubs/EventHubConsumerAsyncClientIntegrationTest.java]: According to your comment, fixed in the new version.

**Comment by benbp** [comment]
@hongli750210 is this ready to merge or are you still adding new commits?

**Comment by v-xuto** [comment]
@benbp Now, this PR is not ready to merge. Because we rerun the event hub weekly pipeline, it failed in `public`, `china` and `UsGov` clouds. We're still investigating.
More details click here: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1507562&view=results

**Comment by v-xuto** [comment]
@benbp Now, this PR is ready to merge, please help merge. The result of the eventhub weekly pipeline is `Pass`. 
More details click here: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1517222&view=results

**Comment by benbp** [comment]
Great, thanks @v-xuto!
